include/maidsafe/crux/detail/multiplexer.hpp

```diff
@@ -113,31 +121,31 @@ class multiplexer : public std::enable_shared_from_this<multiplexer>
 
     template <typename AcceptHandler>
     void process_accept(const boost::system::error_code& error,
-                        socket_base *,
-                        const endpoint_type& remote_endpoint,
                         AcceptHandler&& handler);
 
     boost::asio::io_service& get_io_service() {
         return udp_socket.get_io_service();
     }
 
+    endpoint_type local_loopback_endpoint() const;
+
+    void discard_message();
+
 private:
     next_layer_type udp_socket;
 
     using socket_map = std::map<endpoint_type, socket_base *>;
     socket_map sockets;
 
-    int receive_calls;
+    std::atomic<int> receive_calls;
 
     // FIXME: Move to acceptor class
     // FIXME: Bounded queue with pending accept requests? (like listen() backlog)
     using accept_handler_type = std::function<void (const boost::system::error_code&)>;
-    using accept_input_type = std::tuple<socket_base *, accept_handler_type>;
-    std::queue<std::unique_ptr<accept_input_type>> acceptor_queue;
+    using accept_input_type = std::tuple<acceptor*, socket_base *, accept_handler_type>;
+    std::list<std::unique_ptr<accept_input_type>> acceptor_queue;
```

Comment:
FYI, the original type was `std::queue` so it could more easily be changed into a lock-free data structure (`boost::lockfree::queue`) when we have to do thread-safety.

The overall design goal has been to only use atomic variables or concurrent data structures in `multiplexer` so we could give the basic level of thread-safety guarantees without having different sockets influencing each other too severely.